### PR TITLE
Add charm tracing

### PR DESCRIPTION
**Pull request by mmkay**
This PR adds charm tracing support to mimir-coordinator. It doesn't use TLS on the communication with tempo yet - this will be done later in a follow-up PR.

# How to use tracing

In the model where you have your cos stack, run the following:
```
juju deploy tempo-k8s
juju relate tempo-k8s:grafana-source grafana:grafana-source
<deploy mimir>
juju relate mimir-coordinator-k8s:tracing tempo-k8s:tracing
```

Then in Grafana you can browse your charm traces:
Screenshot from 2024-01-09 17-56-04

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -55,6 +55,12 @@ requires:
     description: |
       Ingress-per-app, to load-balance across multiple units of the coordinator.
 
+  tracing:
+    interface: tracing
+    limit: 1
+    description: |
+      Enables sending traces for charm troubleshooting.
```
Comment (PietroPasotti) [inline metadata.yaml:62]: for now this is fine, but keep in mind some day we might use tracing to monitor more than the charm itself (nginx?)

e.g. send the tracing info forward to the workers so we can send mimir traces to tempo, and the mirmir_worker charm as well.

File: src/charm.py [hunk 2]
```diff
@@ -185,6 +199,14 @@ def _on_nginx_pebble_ready(self, _event: ops.PebbleReadyEvent) -> None:
         self._nginx_container.add_layer("nginx", self.nginx.layer, combine=True)
         self._nginx_container.autostart()
 
+    @property
+    def tempo(self) -> Optional[str]:
```
Comment (PietroPasotti) [inline src/charm.py:203]:
```suggestion
    def tempo_endpoint(self) -> Optional[str]:
```

**Review by PietroPasotti — APPROVED** [review]
some cosmetic thoughts, but looks good overall!
GJ

File: src/charm.py (hunk 2 above)
Comment (mmkay) [reply src/charm.py]: Done.

File: metadata.yaml (hunk 1 above)
Comment (mmkay) [reply metadata.yaml]: True, and that's absolutely wanted for future. Changed the description to be more universal.

**Pull request merged**